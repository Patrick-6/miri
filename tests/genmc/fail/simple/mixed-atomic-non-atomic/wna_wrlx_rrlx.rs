//@compile-flags: -Zmiri-genmc -Zmiri-disable-stacked-borrows
//@revisions: return1234 return42

#![no_main]

use std::ffi::c_void;
use std::sync::atomic::{AtomicU64, Ordering};

use libc::{self, pthread_attr_t, pthread_t};

#[unsafe(no_mangle)]
fn miri_start(_argc: isize, _argv: *const *const u8) -> isize {
    let mut t0: pthread_t = 0;
    let mut t1: pthread_t = 0;

    let attr: *const pthread_attr_t = std::ptr::null();

    let mut x: AtomicU64 = AtomicU64::new(1);
    *x.get_mut() = 42;
    // x.store(42, STORE_ORD);

    let value: *mut c_void = x.as_ptr() as *mut c_void;

    assert!(0 == unsafe { libc::pthread_create(&raw mut t0, attr, read_relaxed, value) });
    assert!(0 == unsafe { libc::pthread_create(&raw mut t1, attr, write_relaxed, value) });

    assert!(0 == unsafe { libc::pthread_join(t0, std::ptr::null_mut()) });
    assert!(0 == unsafe { libc::pthread_join(t1, std::ptr::null_mut()) });

    0
}

extern "C" fn read_relaxed(value: *mut c_void) -> *mut c_void {
    unsafe {
        let x = (value as *const AtomicU64).as_ref().unwrap();
        let val = x.load(Ordering::Relaxed);

        let mut flag = false;
        if cfg!(return1234) && 1234 == val {
            flag = true;
        }
        if cfg!(return42) && 42 == val {
            flag = true;
        }
        if flag {
            std::hint::unreachable_unchecked(); //~ ERROR: entering unreachable code
        }
        std::ptr::null_mut()
    }
}

extern "C" fn write_relaxed(value: *mut c_void) -> *mut c_void {
    unsafe {
        let x = (value as *const AtomicU64).as_ref().unwrap();
        x.store(1234, Ordering::Relaxed);
        std::ptr::null_mut()
    }
}
