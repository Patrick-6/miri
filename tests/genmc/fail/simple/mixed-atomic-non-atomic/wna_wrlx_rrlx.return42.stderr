Running GenMC Verification...
error: Undefined Behavior: entering unreachable code
  --> tests/genmc/fail/simple/mixed-atomic-non-atomic/wna_wrlx_rrlx.rs:LL:CC
   |
LL |             std::hint::unreachable_unchecked();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Undefined Behavior occurred here
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: BACKTRACE on thread `unnamed-ID`:
   = note: inside `read_relaxed` at tests/genmc/fail/simple/mixed-atomic-non-atomic/wna_wrlx_rrlx.rs:LL:CC

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

note: add `-Zmiri-genmc-print-genmc-output` to MIRIFLAGS to see the detailed GenMC error report

error: aborting due to 1 previous error

